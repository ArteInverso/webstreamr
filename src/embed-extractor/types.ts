import { StreamWithMeta } from '../types';

export interface EmbedExtractor {
  readonly id: string;

  readonly label: string;

  readonly extract: (embedUrl: string, language: string) => Promise<StreamWithMeta>;
}
