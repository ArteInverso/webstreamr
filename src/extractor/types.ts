import { Context, CountryCode, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (ctx: Context, url: URL) => boolean;

  readonly normalize: (url: URL) => URL;

  readonly extract: (ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined) => Promise<UrlResult[]>;
}
