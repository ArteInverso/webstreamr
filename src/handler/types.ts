import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface HandleResult {
  countryCode: CountryCode;
  referer?: URL;
  title?: string;
  url: URL;
}

export interface Handler {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  readonly handle: (ctx: Context, type: ContentType, id: Id) => Promise<(HandleResult[])>;
}
