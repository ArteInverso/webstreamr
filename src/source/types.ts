import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface SourceResult {
  countryCode: CountryCode;
  title?: string;
  url: URL;
}

export interface Source {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  handle(ctx: Context, type: ContentType, id: Id): Promise<(SourceResult[])>;
}
