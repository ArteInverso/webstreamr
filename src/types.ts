import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export interface Context {
  id: string;
  ip: string;
  config: Config;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Record<string, string>;

export interface Meta {
  bytes?: number;
  countryCode: string;
  height?: number;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  label: string;
  sourceId: string;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
