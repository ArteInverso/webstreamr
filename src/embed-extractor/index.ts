export * from './EmbedExtractors';
export * from './Dropload';
export * from './SuperVideo';
export * from './types';
