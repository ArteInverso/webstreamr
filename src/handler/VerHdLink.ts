import { ContentType } from 'stremio-addon-sdk';
import * as cheerio from 'cheerio';
import { Handler, HandleResult } from './types';
import { Fetcher, getImdbId, Id } from '../utils';
import { Context, CountryCode } from '../types';

export class VerHdLink implements Handler {
  readonly id = 'verhdlink';

  readonly label = 'VerHdLink';

  readonly contentTypes: ContentType[] = ['movie'];

  readonly countryCodes: CountryCode[] = [CountryCode.es, CountryCode.mx];

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly handle = async (ctx: Context, _type: string, id: Id): Promise<HandleResult[]> => {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const pageUrl = new URL(`https://verhdlink.cam/movie/${imdbId.id}`);
    const html = await this.fetcher.text(ctx, pageUrl);

    const $ = cheerio.load(html);

    return $('._player-mirrors')
      .map((_i, el) => {
        let countryCode: CountryCode;
        if ($(el).hasClass('latino') && 'mx' in ctx.config) {
          countryCode = CountryCode.mx;
        } else if ($(el).hasClass('castellano') && 'es' in ctx.config) {
          countryCode = CountryCode.es;
        } else {
          return [];
        }

        return $('[data-link!=""]', el)
          .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
          .toArray()
          .filter(url => !url.host.match(/verhdlink/))
          .map(url => ({ countryCode, referer: pageUrl, url }));
      }).toArray();
  };
}
