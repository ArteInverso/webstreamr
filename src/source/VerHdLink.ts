import * as cheerio from 'cheerio';
import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getImdbId, Id } from '../utils';
import { Source, SourceResult } from './types';

export class VerHdLink implements Source {
  public readonly id = 'verhdlink';

  public readonly label = 'VerHdLink';

  public readonly contentTypes: ContentType[] = ['movie'];

  public readonly countryCodes: CountryCode[] = [CountryCode.es, CountryCode.mx];

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public async handle(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const pageUrl = new URL(`https://verhdlink.cam/movie/${imdbId.id}`);
    const html = await this.fetcher.text(ctx, pageUrl);

    const $ = cheerio.load(html);

    return $('._player-mirrors')
      .map((_i, el) => {
        let countryCode: CountryCode;
        if ($(el).hasClass('latino') && 'mx' in ctx.config) {
          countryCode = CountryCode.mx;
        } else if ($(el).hasClass('castellano') && 'es' in ctx.config) {
          countryCode = CountryCode.es;
        } else {
          return [];
        }

        return $('[data-link!=""]', el)
          .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
          .toArray()
          .filter(url => !url.host.match(/verhdlink/))
          .map(url => ({ countryCode, referer: pageUrl, url }));
      }).toArray();
  };
}
