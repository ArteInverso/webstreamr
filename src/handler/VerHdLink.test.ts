import { VerHdLink } from './VerHdLink';
import { Fetcher } from '../utils';
import { Dropload, EmbedExtractors, SuperVideo } from '../embed-extractor';
import { Context } from '../types';
jest.mock('../utils/Fetcher');

// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const mostraguarda = new VerHdLink(fetcher, new EmbedExtractors([new Dropload(fetcher), new SuperVideo(fetcher)]));
const ctx: Context = { ip: '127.0.0.1' };

describe('VerHdLink', () => {
  test('does not handle non imdb movies', async () => {
    const streams = await mostraguarda.handle(ctx, 'kitsu:123');

    expect(streams).toHaveLength(0);
  });

  test('handles non-existent movies gracefully', async () => {
    const streams = await mostraguarda.handle(ctx, 'tt12345678');

    expect(streams).toHaveLength(0);
  });

  test('handle titanic', async () => {
    const streams = await mostraguarda.handle(ctx, 'tt0120338');

    expect(streams).toHaveLength(4);
    expect(streams[0]).toStrictEqual({
      url: expect.any(URL),
      label: 'SuperVideo',
      sourceId: 'supervideo_mx',
      height: 556,
      bytes: 1503238553,
      countryCode: 'mx',
    });
    expect(streams[0]?.url.href).toMatch(/^https:\/\/.*?.m3u8/);
    expect(streams[1]).toStrictEqual({
      url: expect.any(URL),
      label: 'Dropload',
      sourceId: 'dropload_mx',
      height: 556,
      bytes: 1503238553,
      countryCode: 'mx',
    });
    expect(streams[1]?.url.href).toMatch(/^https:\/\/.*?.m3u8/);
    expect(streams[2]).toStrictEqual({
      url: expect.any(URL),
      label: 'SuperVideo',
      sourceId: 'supervideo_es',
      height: 544,
      bytes: 1610612736,
      countryCode: 'es',
    });
    expect(streams[2]?.url.href).toMatch(/^https:\/\/.*?.m3u8/);
    expect(streams[3]).toStrictEqual({
      url: expect.any(URL),
      label: 'Dropload',
      sourceId: 'dropload_es',
      height: 544,
      bytes: 1610612736,
      countryCode: 'es',
    });
    expect(streams[3]?.url.href).toMatch(/^https:\/\/.*?.m3u8/);
  });
});
