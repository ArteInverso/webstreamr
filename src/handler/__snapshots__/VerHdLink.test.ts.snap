// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`VerHdLink handle titanic 1`] = `
[
  {
    "countryCode": "mx",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://supervideo.cc/e/g6okbr0kd790",
  },
  {
    "countryCode": "mx",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://dropload.io/embed-jjjx9j5joiz8.html",
  },
  {
    "countryCode": "mx",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://mixdrop.ag/e/vn0wx308fq984q",
  },
  {
    "countryCode": "mx",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://streamtape.com/e/Bjp2vjrdBxsK82",
  },
  {
    "countryCode": "mx",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://dood.to/e/rw7rxdfrbg09",
  },
  {
    "countryCode": "es",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://supervideo.cc/e/qe1fviow8uwy",
  },
  {
    "countryCode": "es",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://dropload.io/embed-ick4ti66vt6s.html",
  },
  {
    "countryCode": "es",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://mixdrop.ag/e/xokzmv61hrwe8o",
  },
  {
    "countryCode": "es",
    "referer": "https://verhdlink.cam/movie/tt0120338",
    "url": "https://dood.to/e/gy8l8mb2i311",
  },
]
`;
